Add StrictInteger type

lanej opened the pull request:
### Summary

```
= String#to_i

...

This method never raises an exception when base is valid.
```

Use `Integer()` to catch invalid integer values and return nil in the error case.  The behavior of the other casting functions and this function to return `nil` in the case of invalid values.

lanej commented:
There is an assumption that string presentations of decimal values parse and truncate the fraction (the `to_i` behavior).  This can be solved with a more complex solution.  lmk what you think.

lanej commented:
I add a commit that uses a regexp to determine decimal values.  A similar method is used in the date type.

rafaelfranca commented:
This is a huge change in behavior. We can't change that unfortunately. If you want a more strict type checking for integer I recommend you to define your own integer type. Or we could have a new integer type with this new behavior but only active with a configuration.

@matthewd do you think I'm being over zealous and changing this behavior is not as big as I think?

matthewd commented:
Use of integers with non-numeric "tails" is fairly widespread for some use cases, I think (mostly when those integers are IDs). I guess we could do it with a deprecation cycle, but I agree it seems likely to affect a good number of people.

A stricter integer type you could opt in to seems like a good middle ground, at least as a first step.

That said, I'm only concerned about the implications of breaking `"1ignore" -> 1`. Changing `"bad1" -> 0` to `"bad1" -> nil` seems like a reasonable (and comparatively safe) improvement to me.

lanej commented:
> That said, I'm only concerned about the implications of breaking "1ignore" -> 1. Changing "bad1" -> 0 to "bad1" -> nil seems like a reasonable (and comparatively safe) improvement to me.

I take your point.  This could break a lot of people.

> A stricter integer type you could opt in to seems like a good middle ground, at least as a first step.

I agree.  I'll get that started.

lanej committed:
add ActiveModel::Type::StrictInteger

Add an integer type that relies upon Integer.() instead of Integer#to_i.
While Type::Integer will return an integer at the beginning of a string
(i.e. "1bad" => 1), this type will return nil unless the provided value
represents a number in it's entirety.

lanej commented:
Added the `StrictInteger` type.  CI failure appears unrelated but I cannot retry the failed job.

lanej commented:
ping for attention.

sgrif commented on activemodel/lib/active_model/type/strict_integer.rb:
```diff
@@ -0,0 +1,19 @@
+          case value
+          when true then 1
+          when false then 0
+          when /\A[\d\.]+\Z/ then value.to_i
```
This regular expression is incorrect. It does not match `-1`, a valid integer. It does match `1.0.0`, which is not a valid integer.

sgrif commented:
I agree that this is a thing that some people may generally find useful. I'm not sure that I see the need for it to be in Rails by default. I think you should try releasing this as a gem, and see if it gains traction. If this becomes a thing that everyone is putting in their Gemfile, let's reconsider adding it to Rails at that point.
